### ISSUE TITLE
Incorrect reporting of "Alias is never used in SELECT statement"

### ISSUE THREAD

**Issue by jamiet-msm** [sqlfluff/sqlfluff#2030]
## Expected Behaviour
sqlfluff shouldn't raise any issues. Its complaining that alias `t` isn't used when in fact it is, its referenced in `TO_JSON_STRING()`

It is valid code 
![image](https://user-images.githubusercontent.com/79862277/144639463-5fc8ea87-47d3-4a70-bf29-509a70319a73.png)

## Observed Behaviour
> L:   4 | P:   6 | L025 | Alias 't' is never used in SELECT statement.

![image](https://user-images.githubusercontent.com/79862277/144609131-7fbccc88-b3aa-4740-b26d-11e70ea71afe.png)

## Steps to Reproduce

put this code into file.sql:

```sql
SELECT TO_JSON_STRING(t) AS json
FROM (
        SELECT 1 AS c
) AS t
```

run `sqlfluff lint --dialect bigquery path/to/file.sql`

## Dialect

bigquery

## Version

sqlfluff, version 0.8.2
Python 3.8.12

## Configuration

n/a

**Comment by tunetheweb** [comment]
That's an interesting bit of SQL there. It's raw to pass a whole result set to a function, which is why I think rule L025 doesn't cater for it.

This works for example and doesn't throw a linting error:

```sql
SELECT TO_JSON_STRING(t.c) AS json
FROM (
        SELECT 1 AS c
) AS t
```

I'm presuming this is a cut down example, and there's a reason you don't do just that, but could you expand upon it a little more so we can decide if this is something L025 should support or not?

**Comment by jpy-git** [comment]
@tunetheweb this is valid syntax (https://roboquery.com/app/syntax-to-json-string-function-bigquery)

TO_JSON_STRING takes in a result set and returns each row in a combined JSON format.

Parsing the query shows that `t` is being interpreted as a column reference (which is correct for most functions) rather than a table reference. Hence why it triggers L025.
![image](https://user-images.githubusercontent.com/80432516/144757265-607ecdfc-ea87-459e-8b19-fb23e74ff8f9.png)

Is there a way we could add to the bigquery grammar so that this specific `TO_JSON_STRING` function parses the `t` as a table reference? Potentially L025 will start behaving once it knows it's looking at a table reference.

**Comment by jamiet-msm** [comment]
> I'm presuming this is a cut down example, and there's a reason you don't do just that, but could you expand upon it a little more so we can decide if this is something L025 should support or not?

To expand a little on what @jpy-git said... simply we have a need to supply a resultset as a JSON document to a downstream consumer of that data (the downstream consumer is a Google Cloud Dataflow job - not that that is significant).

**Comment by barrywhart** [comment]
We've had similar challenges with BigQuery in the past. See the function `sqlfluff.dialects.dialect_bigquery.ColumnReferenceSegment.extract_possible_references()`. It overrides a base-class function in the ANSI dialect to deal with some of the ambiguities of BigQuery syntax. I think we may be able to leverage or tweak this function (or possibly apply a similar technique elsewhere in the BigQuery dialect or in rule L025 to fix this issue.

The above thoughts are pretty vague, just intended to be a general pointer towards an approach, not a definite, specific plan.
